### PR TITLE
improve parsing of !$include's to handle bracket expressions fully

### PR DESCRIPTION
**Pull request by tavisrudd**
Includes some new tests and some random tsformating

### CODE REVIEW

File: src/preprocess/visitor.ts [hunk 1]
```diff
@@ -200,22 +223,50 @@ export class Visitor {
   }
 
   visit$include(node: yaml.$include, path: string, env: Env): AnyButUndefined {
-    if (node.data.indexOf('.') > -1) {
-      const reduce: any = _.reduce; // HACK work around broken lodash typedefs
-      const lookupRes: any = reduce(node.data.split('.'), (result0: any, subKey: string) => {
+    const searchChunks = _parse$includeKeyChunks(node.data, path)
+    const lookupRes: any = _.reduce( // tslint:disable-line:no-any
```
Comment (tuff) [inline src/preprocess/visitor.ts]: Any particular reason to use `_.reduce` here, rather than `searchChunks.reduce(...)`?

File: src/preprocess/visitor.ts [hunk 2]
```diff
@@ -30,6 +30,42 @@ const _isPlainMap = (node: any): node is object =>
...
+      i = 0
+    } else if (remaining[i] === '[') {
+      bracketDepth++
+      i++
+    } else if (remaining[i] === ']') {
+      bracketDepth--
+      i++
+    } else {
+      i++
+    }
```
Comment (tuff) [inline src/preprocess/visitor.ts]: A bit shorter...is it less clear?

```
} else {
      i++

      if (remaining[i] === '[') {
        bracketDepth++
      }
      if (remaining[i] === ']') {
        bracketDepth--
      }
    }
```

File: src/tests/test-visitor.ts [hunk 3]
```diff
@@ -38,23 +38,117 @@ describe('VariablesTrackingVisitor', () => {
...
+      'foo[sub].bar': ['foo[sub]', 'bar'],
+      'foo[sub].bar.baz': ['foo[sub]', 'bar', 'baz'],
+      'foo[sub].bar.baz[sub2]': ['foo[sub]', 'bar', 'baz[sub2]'],
+      'foo.bar[sub].baz': ['foo', 'bar[sub]', 'baz'],
+      'foo[sub0.sub1].bar': ['foo[sub0.sub1]', 'bar'],
+      'foo[sub0][sub1].bar': ['foo[sub0][sub1]', 'bar'],
+      'foo[sub0][sub1].bar[sub3]': ['foo[sub0][sub1]', 'bar[sub3]']
+    }
+    for (const [key, res] of Object.entries(examples)) {
+      expect(_parse$includeKeyChunks(key, 'test'), `input=${key}`).to.deep.equal(res);
```
Comment (tuff) [inline src/tests/test-visitor.ts]: Not sure if I missed something obvious but it took me way too long to discover what the second argument to `expect` is for - a custom error message. TIL.

File: src/preprocess/visitor.ts (hunk 2 above)
Comment (tavisrudd) [reply src/preprocess/visitor.ts]: I prefer the more verbose version.

File: src/preprocess/visitor.ts [hunk 4]
```diff
@@ -200,22 +223,50 @@ export class Visitor {
...
+    for (const part of dynamicKey.trim().split('][')) {
+      // this loop enables handling of successive, unnested dynamic keys
+      // e.g. 'obj[key1][key2]'. See the tests in src/test/test-visitor.ts
+      if (Number.isSafeInteger(parseInt(part, 10))) {
+        newKeyParts.push(part)
       } else {
-        return this.visitNode(lookupRes, path, env);
+        const bracketVal = this.visit$include(new yaml.$include(part), path, env);
+        if (typeof bracketVal === 'number') {
+          newKeyParts.push(bracketVal + '')
```
[the remark this answers is not in the document]
Comment (tavisrudd) [reply src/preprocess/visitor.ts]: changing to that. old habits

File: src/preprocess/visitor.ts (hunk 1 above)
Comment (tavisrudd) [reply src/preprocess/visitor.ts]: old habits. Changing it.

File: src/tests/test-visitor.ts (hunk 3 above)
Comment (tavisrudd) [reply src/tests/test-visitor.ts]: Yep, custom message.

**Comment by jpb** [comment]
I've been playing around with this a bit. There are some weird edge cases that I was able to get into. For example, with:

```
$defs:
  a:
    b:
      c:
        1
  keys:
    keys: keys
    keyA: a
    keyB: b
    keyC: c

foo:
  - !$ a.b.c # => 1
  - !$ keys[keys.keyA][keys.keyB][keys.keyC] # => 1
  - !$ keys[keys[keys.keys]][keys.keyA][keys.keyB][keys.keyC] # => 1

  - !$ keys[keys.keyA] # => { b: c: 1 }
  - !$ keys.a # => { b: c: 1 }

  - !$ keys[keys[keys.keys]] # => keys
  - !$ keys.keys # => keys
```

`keys[keys[keys.keys]]` evaluates to `"keys"` when on it's own, but `{ "keys": { keyA: ...` when chained (eg. `keys[keys[keys.keys]][keys.keyA][keys.keyB][keys.keyC]`). It seems like referential transparency is lost in some cases.

Given the complexity of this code, it makes me wonder if we can offload it to another library. For example, lodash can handle these complex cases. Perhaps we could even drop the use of handlebars by using:

```
_.templateSettings.interpolate = /{{([\s\S]+?)}}/g;
var compiled = _.template('{{ a[keyB].c }}');
console.log(compiled({ keyB: 'b', a: { b: { c: 1 } } }));
```

Lodash correctly errors when using `_.template('{{ keys[keys[keys.keys]][keys.keyA][keys.keyB][keys.keyC] }}');`.

**Comment by tavisrudd** [comment]
I've fixed the issues @jpb found (bad scoping) and experimented with `_.template` and `_.get`. Neither worked well, primarily because of bad error reporting. I think we can follow up with another PR that simplifies this code a bit and improves its own error reporting.

**Comment by tavisrudd** [comment]
Squashing this prior to merge.
